File: pizzashop-api/src/PizzaShop.API/UseCases/GetOrders.cs
﻿using PizzaShop.API.Authentication;
using PizzaShop.API.Domain.Extensions;
using PizzaShop.API.Domain.Interfaces;
using PizzaShop.API.Domain.Models;

namespace PizzaShop.API.UseCases
{
	public sealed class GetOrders(IAuthenticate authenticate, IOrderRepository _orderRepository)
		: UseCaseBase<GetOrdersFiltersDto>(authenticate, _resturantRequired: true)
	{
		public override Task<Result<GetOrdersDto>> Execute(GetOrdersFiltersDto filters)
		{
			var (orders, total) = _orderRepository.GetOrders(
				UserToken.RestaurantId.ToGuid(),
				filters.OrderId,
				filters.Status,
				filters.CustomerName,
				filters.PageIndex);

			var result = new GetOrdersDto();
			foreach (var order in orders)
			{
				result.Orders.Add(new()
				{
					OrderId = order.Id,
					Status = order.Status,
					Total = order.Total,
					CreatedAt = order.CreatedAt,
					CustomerName = order.Customer.Name,
				});
			}

			result.TotalCount = total;
			result.PageIndex = filters.PageIndex;
			result.PerPage = result.Orders.Count > 10 ? 10 : result.Orders.Count;

			return Task.FromResult(new Result<GetOrdersDto> { Data = result });
		}
	}
}

File: pizzashop-web/src/api/get-orders.ts
import { api } from '@lib/axios'

export interface GetOrdersFilters {
  pageIndex?: number | null
  customerName?: string | null
  orderId?: string | null
  status?: string | null
}

export interface GetOrdersResponse {
  orders: {
    orderId: string
    createdAt: string
    status: 'Pending' | 'Canceled' | 'Processing' | 'Delivering' | 'Delivered'
    customerName: string
    total: number
  }[]
  pageIndex: number
  perPage: number
  totalCount: number
}

export async function getOrders({
  pageIndex,
  customerName,
  orderId,
  status,
}: GetOrdersFilters) {
  const response = await api.post<GetOrdersResponse>('/api/v1/orders', {
    pageIndex,
    customerName,
    orderId,
    status,
  })

  try {
    return response.data
  } catch (error) {
    console.log(error)
  }
}
